### PR TITLE
Handle version "0" and break some circular dependencies

### PR DESCRIPTION
**Pull request by ksurent**
Having version "0" in configs and handling it correctly enables `Pakket` to use `Module::CoreList` to skip building core modules, some of which have circular dependencies between each other.
Two notable examples are: `Data::Dumper` and `ExtUtils::MakeMaker`.

### CODE REVIEW

**Comment by ksurent** [comment]
At least one major problem with this approach of using `Module::CoreList` is that it tells you what was bundled in the original perl distribution which is not necessarily what your vendor ships. For instance, in CentOS some core modules are extracted into a separate package (`perl-core`) which renders the `is_core()` check pretty much useless.
I see two obvious possibilities here:
1. Try loading modules (something like `is_core($module) and require_module($module)`)
2. Try finding the module's files on disk (search for `module_notional_filename($module)` in `@INC`)

What do you think?

**Comment by xsawyerx** [comment]
Why not add an option for which perl version you want to use as the base-line for `Module::CoreList`?

    --for 5.24.1

etc.

**Comment by xsawyerx** [comment]
You will need to rebase this. Changes were done yesterday to the scaffolder script. Double check with @mickeyn that he pushed the changes.

**Comment by ksurent** [comment]
I'm not sure how adding `--for` can help with the fact that reality sometimes disagrees with what `Module::CoreList` thinks is right. Here's an example of what I mean. This is what you get on a vanilla CentOS 7:
```
$ cat /etc/redhat-release
CentOS Linux release 7.2.1511 (Core)
$ perl -V:version
version='5.16.3';
$ perl -MExtUtils::MakeMaker -e1
Can't locate ExtUtils/MakeMaker.pm in @INC (@INC contains: /usr/local/lib64/perl5 /usr/local/share/perl5 /usr/lib64/perl5/vendor_perl /usr/share/perl5/vendor_perl /usr/lib64/perl5 /usr/share/perl5 .).
BEGIN failed--compilation aborted.
```

While `Module::CoreList` says this:
```
$ corelist ExtUtils::MakeMaker

Data for 2016-05-20
ExtUtils::MakeMaker was first released with perl 5
```

The reason for that is that in CO7 a bunch of core modules are extracted in _a separate package which is not installed by default_.

Further on, I think the version option is unnecessary because we can and arguably even should find it automatically by introspecting perl that we're building against.

Having said all this, perhaps another possibility could be to just make this the user's responsibility. That is, document that for Pakket to work properly it needs perl and all of its core modules. Even if the modules need to be installed separately.

**Comment by xsawyerx** [comment]
I see what you mean now.

In this sense, we can do the following:

1. Check whether a module is available on the perl itself. We're going to run `perl` for this anyway, we can run it to check whether a module is available.
2. Possibly easier, checking for dual-life modules. Would that make it easier?

File: lib/Pakket/Builder.pm [hunk 1]
```diff
@@ -607,8 +608,17 @@ $build_perl_version = Perl::Version->new($build_perl_version)->numify;
...
 
-    return Module::CoreList::is_core( $dep, $dep_version,
+
+    my $core = Module::CoreList::is_core( $dep, $dep_version,
         $build_perl_version );
+
+    my $available
+        = $core && Module::Runtime::use_module( $dep, $dep_version );
+
+    log_debug {"$dep $dep_version: is core: $_[0]; available: $_[1]"}
```
Comment (xsawyerx) [inline lib/Pakket/Builder.pm]: `$log->debug()`, no?

**Comment by xsawyerx** [comment]
:+1: to recent changes.

File: lib/Pakket/Builder.pm (hunk 1 above)
Comment (ksurent) [inline lib/Pakket/Builder.pm]: Yeah, I need to rebase the whole branch. In its current state it's still using `Log::Contextual`.

**Pull request closed without merging**